mlflow/utils/conda.py

```diff
@@ -133,10 +181,29 @@ def get_or_create_conda_env(conda_env_path, env_id=None, capture_output=False):
             )
         )
 
-    env_names = _list_conda_environments()
+    conda_extra_envs = _get_conda_extra_envs(env_root_dir)
+
     project_env_name = _get_conda_env_name(conda_env_path, env_id)
-    if project_env_name not in env_names:
-        _logger.info("=== Creating conda environment %s ===", project_env_name)
+    if env_root_dir is not None:
+        # Append a suffix "-isolated" because if a conda env name exist in
+        # default conda env root dir, then if we set new "CONDA_ENVS_PATH" and run "conda env create"
+        # with the same conda env name, "CondaValueError: prefix already exists" error will
+        # be raised.
+        project_env_name = project_env_name + "-isolated"
+
+    if env_root_dir is not None:
+        # Generate the full path of the conda env
+        # Checking whether env was created using full path instead of env name,
+        # otherwise it might match existing env under other different env_root_dir.
+        project_env_path = os.path.normpath(
+            os.path.join(env_root_dir, _CONDA_ENVS_DIR, project_env_name)
+        )
+    else:
+        project_env_path = project_env_name
```

Comment:
Can we process everything using a path? Using both name and path is a bit confusing and complicates the code.

Reply:
If so , the default case (i.e. without CONDA_ENVS_PATH setting), we need run `conda info --json` and parse the result "envs_dirs" entry (It might includes multiple dirs and we need to check the first dir which is writable). The logic is also a bit complex.

Reply:
Can we instead do `conda info --envs --json`, look at the entries under the `"envs"` key, and then find the entry whose `os.path.basename() == project_env_name`? 

Here's an example of the `envs` key content:

```
"envs": [
    "/Users/corey.zumar/opt/anaconda3",
    "/Users/corey.zumar/opt/anaconda3/envs/base2",
    "/Users/corey.zumar/opt/anaconda3/envs/foo",
    "/Users/corey.zumar/opt/anaconda3/envs/javapydiff",
    "/Users/corey.zumar/opt/anaconda3/envs/mlflow-1b55d4c51bfa467f642492d4eb4474cda2c85ee1",
    "/Users/corey.zumar/opt/anaconda3/envs/mlflow-2db1ebf9ea6f978512f24e8516507b38fb7ad1fc",
    "/Users/corey.zumar/opt/anaconda3/envs/mlflow-515e205e1ab623cdd02144de8e4d25085796e193",
    "/Users/corey.zumar/opt/anaconda3/envs/mlflow-54ade87f6825f3ab12eac103b777d6cda0d13852",
    "/Users/corey.zumar/opt/anaconda3/envs/mlflow-795b1deee44b2fe066ea40822019bf3c00b1cfb0",
    "/Users/corey.zumar/opt/anaconda3/envs/mlflow-7ef5a6c33fa6db6e310779594c9919d72af7ea28",
    "/Users/corey.zumar/opt/anaconda3/envs/mlflow-9acaa695f270d7ef672ecf4e5b6beeb77cd423b1",
    "/Users/corey.zumar/opt/anaconda3/envs/mlflow-a86033905519765fe1d87442150b5284ca965398",
    "/Users/corey.zumar/opt/anaconda3/envs/mlflow-af9fca66a859c021675ec43c9c75c12080cfc17e",
    "/Users/corey.zumar/opt/anaconda3/envs/mlflow-b6ecf73bc09f0108678380d09092e88bc31bbf3e",
    "/Users/corey.zumar/opt/anaconda3/envs/mlflow-cbb4f40ff67c4094fa0737599b864e21d2ff364c",
    "/Users/corey.zumar/opt/anaconda3/envs/mlflow-d080e661670d8d0df6823d2cc5af762cf782e60b",
    "/Users/corey.zumar/opt/anaconda3/envs/mlflow-e9c06b6c2e9a28c4146182cd65ccda34e27be6af",
    "/Users/corey.zumar/opt/anaconda3/envs/mlflow-env",
    "/Users/corey.zumar/opt/anaconda3/envs/mlflow-env2",
    "/Users/corey.zumar/opt/anaconda3/envs/mlflowskinny",
    "/Users/corey.zumar/opt/anaconda3/envs/pipdeps",
    "/Users/corey.zumar/opt/anaconda3/envs/py36_failtest",
    "/Users/corey.zumar/opt/anaconda3/envs/python=3.8",
    "/Users/corey.zumar/opt/anaconda3/envs/skinny",
    "/Users/corey.zumar/opt/anaconda3/envs/skinny2",
    "/Users/corey.zumar/opt/anaconda3/envs/tf1",
    "/Users/corey.zumar/opt/anaconda3/envs/tf204",
    "/Users/corey.zumar/opt/anaconda3/envs/tf2042",
    "/Users/corey.zumar/opt/anaconda3/envs/tf24",
    "/Users/corey.zumar/opt/anaconda3/envs/tf25",
    "/Users/corey.zumar/opt/anaconda3/envs/xgbrecent"
  ],
```

This way, like @harupy suggested, we don't need any special casing.

Reply:
@WeichenXu123 Is it true that conda environment paths must have unique basenames?

E.g. the following would be invalid?

```
"/Users/corey.zumar/opt/anaconda3/envs/tf24",
"/Users/corey.zumar/tmp/my_other_envs/tf24",
```

This seems to be implied by your comment here: https://github.com/mlflow/mlflow/pull/5561/files#diff-69f4364af03d692af1d39ecf262500cf91306991c5a2dc52dde9844f1a3f7f1bR188-R191.

If this is the case, can't we list the environment basenames (i.e. take `os.path.basename()` for each of the outputs of `_list_conda_environments()`) and just check if `project_env_name` is in the list?

Reply:
> Is it true that conda environment paths must have unique basenames?

That's the hidden bug I found yesterday.

(1) Between default conda env root path and custom conda env root path, basenames  must be unique.
(2) Between multiple custom conda env root paths, basenames can be the same.

e.g. You can try run following commands one by one

```
conda env create -f conda.yaml -n name_1
CONDA_ENVS_PATH=/tmp/a1 conda env create -f conda.yaml -n name_1  # conflict name failure
CONDA_ENVS_PATH=/tmp/a1 conda env create -f conda.yaml -n name_1-isolated # works
CONDA_ENVS_PATH=/tmp/a2 conda env create -f conda.yaml -n name_1-isolated # still works, although name conflict with the one in `/tmp/a1`

conda env list --json  # NOTE: it will list all the envs (under all custom root paths) created in above....
```

Reply:
But, I adopt your suggestion:
https://github.com/mlflow/mlflow/pull/5561/files#r844813894

> Instead of appending isolated, hash the environment root directory path and append it for uniqueness

This address the issue in a simpler way